clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1120,17 +1123,28 @@ Node selectReadReplica(TopicPartition partition, Node leaderReplica, long curren
         }
     }
 
+    /**
+     * If we have seen new metadata (as tracked by {@link org.apache.kafka.clients.Metadata#updateVersion()}), then
+     * we should check that all of the assignments have a valid position.
+     */
+    private void maybeValidateAssignments() {
```

Comment:
The name here through me off because of its generality. I think we need to work metadata into the name to make it clearer. Maybe `validatePositionsOnMetadataChange` or `validatePositionsForLeaderChanges`.